src/main/java/carpet/script/value/ScreenHandlerValue.java

```diff
@@ -0,0 +1,380 @@
+            return screen;
+        }, this.name);
+    }
+
+    private void openScreen(NamedScreenHandlerFactory factory) {
+        if(this.player == null) return;
+        OptionalInt optionalSyncId = this.player.openHandledScreen(factory);
+        if(optionalSyncId.isPresent() && this.player.currentScreenHandler.syncId == optionalSyncId.getAsInt()) {
+            this.screenHandler = this.player.currentScreenHandler;
+        }
+    }
+
+    public void close() {
+        if(this.player.currentScreenHandler != this.player.playerScreenHandler) {
+            //prevent recursion when closing screen in closing screen callback by doing this before triggering event
+            this.player.currentScreenHandler = this.player.playerScreenHandler;
+            this.player.closeHandledScreen();
+            screenHandler = null;
+        }
+    }
+
+    public boolean isOpen() {
+        if(this.screenHandler == null) {
+            return false;
+        }
+        if(this.player.currentScreenHandler.equals(this.screenHandler)) {
+            return true;
+        }
+        this.screenHandler = null;
+        return false;
+    }
+
+
+    private boolean callListener(PlayerEntity player, String action, int index, int button) {
+        Value playerValue = EntityValue.of(player);
+        Value actionValue = StringValue.of(action);
+        Value indexValue = NumericValue.of(index);
+        Value buttonValue = NumericValue.of(button);
+        List<Value> args = Arrays.asList(this,playerValue,actionValue,indexValue,buttonValue);
+        LazyValue cancel = this.callback.callInContext(this.callbackContext, Context.VOID, args);
+        Value cancelValue = cancel.evalValue(this.callbackContext);
+        return cancelValue.getString().equals("cancel");
+    }
+
+    private void addListenerCallback(ScreenHandler screenHandler) {
+        if(this.callback == null) return;
+
+        screenHandler.addListener(new ScarpetScreenHandlerListener() {
+            @Override
+            public boolean onSlotClick(PlayerEntity player, SlotActionType actionType, int slot, int button) {
+                return ScreenHandlerValue.this.callListener(player,actionTypeToString(actionType),slot,button);
+            }
+            @Override
+            public boolean onButtonClick(PlayerEntity player, int button) {
+                return ScreenHandlerValue.this.callListener(player,"button",button,0);
+            }
+            @Override
+            public void onClose(PlayerEntity player) {
+                ScreenHandlerValue.this.callListener(player,"close",0,0);
+            }
+            @Override
+            public void onSlotUpdate(ScreenHandler handler, int slotId, ItemStack stack) {}
+            @Override
+            public void onPropertyUpdate(ScreenHandler handler, int property, int value) {}
+        });
+    }
+
+    // copied from inventory_* functions, but modified for a list of slots
```

Comment:
I smell that all 'inventory_***' style calls could be nicely abstracted here, but I might be wrong

Reply:
Yeah thats what i thought too, but didn't want to touch the existing stuff.

My approach would be to let `InventoryLocator` accept screen values, and have different subclasses of it. Then return either in instance with Methods for Inventories or for screens.

If you are fine with me adding that, i would be up for doing it. If you rather do it yourself, thats fine too.

Reply:
Alright so my plan is: Have an abstract class `ItemLocator` with abstract methods for the `inventory_*` functions. And then have a `InventoryItemLocator`, which gets constructed just like before, and has the implementation of all `inventory_*` functions as they are now. The `ScreenItemLocator` gets constructed from a screen handler, and would have the implementation for screen values. It would probably be possible to implement the functionality of inventory_find right in the abstract class, and only override methods for getting item stacks or the inventory size.
@gnembon should I do this within this PR, or make a seperate PR for it?